### ISSUE TITLE
Determine how to convert OpenAPI spec files into public docs for CRDB V2 API

### ISSUE THREAD

**Issue by mikeCRL** [cockroachdb/docs#10231]
## Background

Our new CockroachDB API V2 (for v21.1) and the forthcoming CockroachCloud API (for v21.2) are both RESTful APIs for which we'll want to leverage an OpenAPI spec document to generate documentation.

For the CRDB API, @itsbilal has embedded the content in code comments, which then get built into the OpenAPI spec file, as needed, during `make generate`. You can view the details in this PR: https://github.com/cockroachdb/cockroach/pull/62560. The initial spec file is now in the `cockroach` repo at [/docs/generated/swagger/spec.json](https://github.com/cockroachdb/cockroach/blob/master/docs/generated/swagger/spec.json).

He also built into CRDB a script that generates API docs from this—and serves them locally—using [redoc](https://github.com/Redocly/redoc). I have also created similar views that are public and easier to access when testing alternate solutions:

* `widdershins` & [fork of] `slate` https://slatetest.netlify.app/
* `jekyll-openapi`: https://apidocdev.netlify.app/
* `Readme.com`: https://apidocdev.readme.io/reference

## Problem to solve

Decide on a way to render an OpenAPI spec document on the doc site. Determine if this will be the same method for both APIs. (Note: @petermattis confirmed that we will plan to use an OpenAPI spec file for the CockroachCloud API, as well.) Or perhaps an interim solution for the CRDB API for v21.1 and a refined, matching solution for both APIs after that.

If there is insufficient time/bandwidth to develop a longer term solution by the 21.1 release, we'll need an interim solution for publishing these docs. For example, one-off use of a static site generator, some other form of generated docs that we'll embed/include in docs pages, manually authoring all the docs content for this release, or linking out to a hosted service.

**Edit:** The [updated proposal](https://docs.google.com/document/d/1mB94iS_Aqm8p4y98x703lzGDxN8awQH31-FA3PmF5-g/edit) is to focus on a solution exclusive to the CRDB V2 API for the v21.1 release using a static site generator.

### Potential solutions

   - There are various [tools for building docs from OpenAPI](https://openapi.tools/#documentation); most build a whole site, with a layout geared toward API docs. In our case, it might be best to adapt one of these to match the design of our doc site and include those pages as a separate path when we build the doc site. One option worth considering is https://github.com/robertlove/jekyll-openapi given that our existing site is built using Jekyll.
   - Alternatively we could have Jekyll parse the spec directly and include its content in 'normal' doc pages. See Option 3 here https://idratherbewriting.com/learnapidoc/pubapis_combine_swagger_and_guide.html. Ignore the old links out from there; the updated ones are: [Example result](https://peterhend.github.io/documentation-theme-jekyll/api_post_pet.html) and [source](https://github.com/peterhend/documentation-theme-jekyll).
   - For either of these APIs, a third route would be to use a hosted service like `readme.com`.

**Edit:** The new [proposed solution](https://docs.google.com/document/d/1mB94iS_Aqm8p4y98x703lzGDxN8awQH31-FA3PmF5-g/edit#heading=h.v4d0pwtnh23t) is to use [Widdershins](https://github.com/Mermade/widdershins) and ~[Dracula fork of](https://github.com/dracula/slate)~ [Slate](https://github.com/slatedocs/slate) (because it's well ahead of that other fork) to build a Middleman-generated static site based on our spec.

### Solution for CRDB v21.1

Edit: This solution is [on hold](https://github.com/cockroachdb/docs/issues/10231#issuecomment-831638620)

- Use Widdershins to convert the OpenAPI spec to Markdown.
- Use Slate to convert the Markdown to a single-page Middleman static site.
- Include both these process in the same CI/CD flow.
- Include that site under the docs site base path, tentatively at https://cockroachlabs.com/docs/api/cluster/v2.0/reference. @mikeCRL's first choice was https://cockroachlabs.com/docs/v21.1/api/cluster/reference/ but @yzdocs advised that a redirect already exists for the v21.1 path pointing to the folder in the docs repo, so this will not work. A different top-level path is needed.

#### Tasks:

- [x] Create a [build script](https://github.com/mikeCRL/apitools/blob/main/build.sh) to combine both tools, turning the latest spec file into docs. @mikeCRL
- [x] Create a separate repo to store the necessarily files. https://github.com/cockroachdb/api-docs @mikeCRL
- [x] Set up a [Netlify CD process](https://github.com/cockroachdb/api-docs/commit/62ba29bdb8ed3b37d7e64a4d60e85a0356c7eeab). @yzdocs
- [ ] Engage Design on recommending some style improvements prior to the v21.1 release. @jseldess or @mikeCRL.
   - We can share the temporary URL https://cockroachdb-api-docs.netlify.app which is currently generated by https://github.com/cockroachdb/api-docs.
- [ ] Engage Marketing to help implement style improvements. @jseldess or @yzdocs.
- [ ] Optional: Redirect future reference docs path to the Netlify domain. https://github.com/cockroachdb/docs/pull/10451 @yzdocs 
- [ ] Begin to deploy the api-docs repo to `/docs/api/cluster/v2.0/reference` directly. @yzdocs.
       - Note that per [Netlify](https://docs.netlify.com/routing/redirects/rewrites-proxies/): "By default, we limit internal rewrites to one “hop” — you cannot proxy from Netlify SiteA to Netlify SiteB to Netlify SiteC in a single request." We already used up one internal rewrite from /docs to marketing site, so the API docs will resolve to https://cockroachdb-api-docs.netlify.com in browser URL bars. Is that fine? Otherwise, we can contact Netlify support as the Netlify docs suggest.

**Comment by mikeCRL** [comment]
Can I ask you to please review this @yingzhuchin, @jseldess, @itsbilal, @piyush-singh, @johnrk and let me know if you have any thoughts on the approach? Is this work possible for v21.1 and does @yingzhuchin have the bandwidth for it?

**Comment by jseldess** [comment]
@mikeCRL, for 21.1, does [this doc](https://docs.google.com/document/d/1mB94iS_Aqm8p4y98x703lzGDxN8awQH31-FA3PmF5-g/edit) supersede this issue?

**Comment by mikeCRL** [comment]
@jseldess Thanks. I have just updated this issue to reflect, at a high-level, what's proposed in the doc and link out to it for more detail. We can continue to use this issue to track and plan work, if that works for you and @yingzhuchin.

**Comment by mikeCRL** [comment]
@jseldess, @yzdocs: @itsbilal and I found a problem with the output of this toolchain today. Despite his help troubleshooting my changes to the spec to add custom examples, they are not appearing in the rendered output. We believe the best option here is to replace this solution with ReDoc. We confirmed that it resolves the problem with examples, and _also_ drastiaclly improves how JSON response details are displayed: as an expandable hierarchy within a single section, rather than linking to separate sections. I'm going to try to spin up and share a demo site shortly and will discuss this further with you both tomorrow.

**Comment by mikeCRL** [comment]
@yzdocs Regarding
> We already used up one internal rewrite from /docs to marketing site, so the API docs will resolve to https://cockroachdb-api-docs.netlify.com in browser URL bars. Is that fine? Otherwise, we can contact Netlify support as the Netlify docs suggest.

I think we should find a way to bring this under a CRL domain. I'd contact support to run it by them and get an answer as to whether this is possible, and how much work to implement. If all they need to do is manually approve, that's not so bad. If it's more complicated, could we just create a separate base path like `cockroachlabs.com/apidocs/` (and use `cockroachlabs.com/apidocs/cluster/v2.0/`)?

**Comment by mikeCRL** [comment]
TL;DR I was able to build Redoc manually https://crdb-redoc.netlify.app and also test their SaaS service https://cockroachdb.redoc.ly . The manual builds should be fine, and preferred, as they let us tailor styles, and don't have a cap on views before bumping to an expensive tier. Will follow up with @yzdocs and @itsbilal after my call with Jesse this morning.

--- 

When building the site with Redoc locally...

```
npm install redoc-cli
npx redoc-cli bundle spec.json
```

...I realized that it wasn't generating and displaying curl requests. There's a tool called [oas3-api-snippet-enricher](https://github.com/cdwv/oas3-api-snippet-enricher) to add these to the spec in a way that Redoc will pick up, but it requires an OpenAPI v3.0 spec. Trying a conversion tool: https://github.com/Mermade/oas-kit/blob/master/packages/swagger2openapi/README.md :

```
npm install -g swagger2openapi 
npx swagger2openapi spec.json > spec_30.json
```

then 

```
npm install snippet-enricher-cli
npx snippet-enricher-cli --input=spec_30.json > spec_30_enriched.json
```

The re-building with Redoc using that new spec file.

Site looks good https://crdb-redoc.netlify.app.

Need to find a way (via the Redoc docs) to tailor which languages show up and what they're called.

Before testing all of the above, the missing curl led me to try the SaaS service Redoc.ly, which owns Redoc. (Their interface and features are very similar to Netlify's, and they offer a WYSIWYG interface for refining Redoc settings, many of which I also see in Redoc itself.) Here' request samples are automatically created without an extra tool. (curl and other languages of your choosing). I have this live now at https://cockroachdb.redoc.ly . It would cost $79/month and have a cap of 1,000 pageviews a month. Requiring more pageviews would bump you to the $360/month level. I also had issues with the way the site displayed after modifying the Google Fonts available. Due to the cap on views and the font issue, it's worth an attempt to get the static site working ourselves.

**Comment by mikeCRL** [comment]
@yzdocs Synced with Jesse:
* OK to switch to building the Redoc site instead as long as we're still meeting users' content needs as agreed by PM/eng - which is the case.
* **No need** to engage Design but we can make some style improvements to the extent we're able.
* Worth asking Netlify support about the additional redirect. If not possible:
   * we can use a different Netlify subdomain (maybe one of the ones you or I already created).
   * Not necessary but as a slight improvement, time permitting or even after the release, we could request a subdomain like api.cockroachlabs.com and point it to the Netlify site [via a CNAME](https://docs.netlify.com/domains-https/custom-domains/configure-external-dns/#configure-a-subdomain), instead.

To set up https://crdb-redoc.netlify.app I just renamed the Redoc-generated HTML file to index.html and manually uploaded it, but we should probably repurpose https://github.com/cockroachdb/api-docs for creating and deploying the new site instead. Is that something you can take a look at?

Also: I'd like to implement some Redoc site [configuration settings](https://github.com/Redocly/redoc#configuration) - I doubt we'll need any swagger vendor extensions but definitely some of the [`<redoc> options object`](https://github.com/Redocly/redoc#redoc-options-object) items, but I'm not yet sure how or where these are included. If it's obvious to you, please let me know!

**Comment by yzdocs** [comment]
> configuration settings

When using the CLI, looks like you can simply pass in your desired config options: https://github.com/Redocly/redoc/blob/master/cli/README.md

**Comment by mikeCRL** [comment]
Reopening while final tweaks are made prior to v21.1 release, then moving outstanding items to a new issue.
